### ISSUE TITLE
[Package Issue]: MSYS2.MSYS2

### ISSUE THREAD

**Issue by HunterZ** [microsoft/winget-pkgs#127282]
### Please confirm these before moving forward

- [X] I have searched for my issue and not found a work-in-progress/duplicate/resolved issue.
- [X] I have not been informed if the issue is resolved in a preview version of the winget client.

### Category of the issue

Installation issue.

### Brief description of your issue

Upgrade from 20230718 to 20231026 fails silently with exit code 1.

Is MSYS2 not meant to be upgraded in this way, since it has an internal pacman-based updater?

### Steps to reproduce

Install 20230718. Attempt to upgrade to 20231026.

### Actual behavior

Upgrade fails with exit code 1 and no additional info.

### Expected behavior

Successful update, or not listed as an available update if this is not a supported means of updating an existing MSYS2 installation.

### Environment

```raw
Windows Package Manager v1.6.3133
Copyright (c) Microsoft Corporation. All rights reserved.

Windows: Windows.Desktop v10.0.22621.2715
System Architecture: X64
Package: Microsoft.DesktopAppInstaller v1.21.3133.0

Winget Directories
-------------------------------------------------------------------------------------------------------------------------------
Logs                               %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\DiagOutputDir
User Settings                      %LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\settings.json
Portable Links Directory (User)    %LOCALAPPDATA%\Microsoft\WinGet\Links
Portable Links Directory (Machine) C:\Program Files\WinGet\Links
Portable Package Root (User)       %LOCALAPPDATA%\Microsoft\WinGet\Packages
Portable Package Root              C:\Program Files\WinGet\Packages
Portable Package Root (x86)        C:\Program Files (x86)\WinGet\Packages
Installer Downloads                %USERPROFILE%\Downloads

Links
---------------------------------------------------------------------------
Privacy Statement   https://aka.ms/winget-privacy
License Agreement   https://aka.ms/winget-license
Third Party Notices https://aka.ms/winget-3rdPartyNotice
Homepage            https://aka.ms/winget
Windows Store Terms https://www.microsoft.com/en-us/storedocs/terms-of-sale

Admin Setting                             State
--------------------------------------------------
LocalManifestFiles                        Disabled
BypassCertificatePinningForMicrosoftStore Disabled
InstallerHashOverride                     Disabled
LocalArchiveMalwareScanOverride           Disabled
```

### Screenshots and Logs

InstallationLog.txt says:

`TargetDirectoryInUse : Error : The directory you selected already exists and contains an installation. Choose a different target for installation.`

**Comment by Smithjuanita** [comment]
Thanks for your excellent service 👏

**Comment by SpecterShell** [comment]
> Is MSYS2 not meant to be upgraded in this way, since it has an internal pacman-based updater?
Right

I tried to prevent it from being updated by WinGet by setting `RequireExplicitUpgrade: true` in the manifest but sometimes it doesn't work. You can try `winget pin add MSYS2.MSYS2` to pin the package.

**Comment by tillig** [comment]
Just ran into this myself. Is exit code 1 unique to the "the directory already exists and contains an installation" error? I wonder if [the `fileInUse` expected error code tied to exit code 1](https://github.com/microsoft/winget-cli/blob/7b82c8593a21e35a597ed1a3f76be50d645d4b11/schemas/JSON/manifests/v1.7.0/manifest.installer.1.7.0.json#L227) might provide something slightly more user friendly. It currently just fails and doesn't provide a pointer to the install log or anything.

```text
 pwsh> winget upgrade --all
Name  Id          Version  Available Source
-------------------------------------------
MSYS2 MSYS2.MSYS2 20230526 20231026  winget
1 upgrades available.

Installing dependencies:
(1/1) Found MSYS2 [MSYS2.MSYS2] Version 20231026
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Successfully verified installer hash
Starting package install...
Installer failed with exit code: 1

 pwsh> 
```

**Comment by SpecterShell** [comment]
> Just ran into this myself. Is exit code 1 unique to the "the directory already exists and contains an installation" error? I wonder if [the `fileInUse` expected error code tied to exit code 1](https://github.com/microsoft/winget-cli/blob/7b82c8593a21e35a597ed1a3f76be50d645d4b11/schemas/JSON/manifests/v1.7.0/manifest.installer.1.7.0.json#L227) might provide something slightly more user friendly. It currently just fails and doesn't provide a pointer to the install log or anything.
> 
> ```
>  pwsh> winget upgrade --all
> Name  Id          Version  Available Source
> -------------------------------------------
> MSYS2 MSYS2.MSYS2 20230526 20231026  winget
> 1 upgrades available.
> 
> Installing dependencies:
> (1/1) Found MSYS2 [MSYS2.MSYS2] Version 20231026
> This application is licensed to you by its owner.
> Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
> Successfully verified installer hash
> Starting package install...
> Installer failed with exit code: 1
> 
>  pwsh> 
> ```

No. The exit code 1 indicates a general installation failure according to [the docs of Qt Installer Framework](https://doc.qt.io/qtinstallerframework/qinstaller-packagemanagercore.html#Status-enum), which is the installer technology used by MSYS2 installer.

**Comment by SpecterShell** [comment]
> I think MSYS2 does not need upgrades as an installer. You can upgrade it via `pacman -Suy` inside its shell.
> 
> So does it mean we don't need this as a winget application?

You can still do first-time install, or even install another copy of MSYS2 if needed. 🤔

**Comment by SpecterShell** [comment]
> > You can still do first-time install, or even install another copy of MSYS2 in a different directory if needed. 🤔
> 
> Then you have a nagging winget upgrade notifier in your hand, which doesn't work.

Not THEN. WinGet will always notify you as long as it detects a copy of MSYS2 from the registry.

I am thinking about modifying the package so it can still install new copies but never detect and update them. `RequireExplicitUpgrade: true` appears to be working only when it was installed through WinGet, and WinGet will still try to update them if it is installed by downloading and running the installer manually.

(AFAIK the counterpart on Chocolatey upgrades by invoking pacman directly, which can not be implemented here at the moment)

**Comment by SpecterShell** [comment]
> You can also install/extract MSYS2 from tar file.

Currently WinGet doesn't support extracting tar file (as it doesn't rely on 7z for extracting archives). See <https://github.com/microsoft/winget-cli/issues/2899>

[The counterpart on Scoop](https://github.com/ScoopInstaller/Main/blob/master/bucket/msys2.json) installs in this way though but I don't know how they tackle the upgrading issue as it persists only the home directory.

**Comment by mdanish-kh** [comment]
> I am thinking about modifying the package so it can still install new copies but never detect and update them. Currently I am using RequireExplicitUpgrade: true but it appears to be working only when it was installed through WinGet, and WinGet will still try to update them if it is installed by downloading and running the installer manually.

When we have the 1.6.0 schema out, maybe `UpgradeBehavior: deny` can help with this one... :eyes:

**Comment by Aalivexy** [comment]
any update?

**Comment by eabase** [comment]
Just got this issue. 
Is there a way to pin a previously installed package to get rid of the winget update notification?

**Comment by eabase** [comment]
I mangled to pin the current installed version with: 
```bash
# winget pin add -q 'MSYS2.MSYS2'

Found MSYS2 Installer [MSYS2.MSYS2]
Pin added successfully
```

**Comment by Trenly** [comment]
Upgrade behavior has been set to deny, and this package requires explicit upgrade. Therefore, it should be skipped by `winget upgrade --all` and should be unable to be upgraded using `winget upgrade msys2.msys2`. As mentioned above, the upgrade notification can be suppressed by pinning the package.

###### Close with reason: Completed;
